sympy/core/exprtools.py

```diff
@@ -1414,156 +1290,549 @@ def factor_nc(expr):
+        return _NCPoly(new_rep, self.generators)
+
+    def as_expr(self):
+        """Convert to Expr"""
+        return Add(*[monom.as_expr() for monom in self.monomials()])
+        
+    def monomials(self, degree=None):
+        """Return iterator of all monomials (with specified degree)"""
+        if degree is None:
+            for monomials in self.rep.values():
+                for monomial in monomials.values():
+                    yield monomial
+        else:
+            for monomial in self.rep.get(degree, {}).values():
+                yield monomial
+
+    def extract_coeff_factor(self, degree=None):
+        from sympy.simplify.radsimp import fraction
+        from sympy.polys.polytools import gcd, lcm
+        numerators = []
+        denominators = []
+        for monom in self.monomials(degree):
+            num, denom = fraction(monom.coeff.cancel())
+            numerators.append(num)
+            denominators.append(denom)
+        return gcd(numerators) / lcm(denominators)
+ 
+    def add(self, monom):
+        """Add a monomial to this polynomial, updating in-place"""
+        degree = monom.degree
+        term = monom.term
+        if degree > self.degree:
+            self.degree = degree
+
+        if degree not in self.rep:
+            self.rep[degree] = {}
+
+        if term in self.rep[degree]:
+            new_monom = self.rep[degree][term] + monom
+
+            if new_monom.coeff == 0:
+                del self.rep[degree][term]
+                if len(self.rep[degree]) == 0:
+                    del self.rep[degree]
+                    if degree == self.degree:
+                        self._eval_degree()
+            else:
+                self.rep[degree][term] = new_monom
+        else:
+            self.rep[degree][term] = monom
+
+    def subtract(self, monom):
+        """Subtract a monomial from this polynomial, updating in-place"""
+        return self.add(monom * -1)
+
+
+    def _factor_homogeneous(self, ldegree):
+        """Factor the maximum-degree terms into two polynomials
+        of degree ldegree and self.degree - ldegree, if possible.
+
+        Return (None, None) if no such factorization exists.
+
+        Implementation based on Algorithm 1 from https://arxiv.org/abs/1002.3180
+        """
+        rdegree = self.degree - ldegree
+        terms = list(self.monomials(self.degree))
+        Ghat, Hhat = terms[0].split(ldegree)
+        G = _NCPoly(None, self.generators)
+        H = _NCPoly(None, self.generators)
+        G.add(Ghat)
+        H.add(Hhat)
+        for monomial in terms[1:]:
+            R = monomial.ldivide(Ghat)
+            if R is None:
+                L = monomial.rdivide(Hhat)
+                if L is not None:
+                    G.add(L)
+            else:
+                H.add(R)
+        if len(G) * len(H) != len(terms):
+            return None, None
+        GH = G * H
+        if GH.as_expr() == Add(*[term.as_expr() for term in terms]):
```

Comment:
I think that you want to imply `G*H == self`, 
however, I could have better come up with an idea to test multiply with NCPoly without help from SymPy

Reply:
Ah, yeah, I think I wrote this just to be able to test whether the function was working as quickly as possible, and then forgot to update.